File: frontend/public/js/permissions.js
/**
 * Mirrors backend/permissions.py. Kept in sync manually for now (there's
 * no shared package between the Flask backend and this plain-JS frontend).
 * If you add/change a permission on the backend, update it here too --
 * otherwise the UI and API will disagree about who can do what.
 */

const PERMISSIONS = {
  super_admin: new Set(['*']),
  librarian: new Set([
    'books:read', 'books:create', 'books:update', 'books:delete',
    'members:read', 'members:create', 'members:update', 'members:delete',
    'circulation:issue', 'circulation:return', 'circulation:renew',
    'circulation:reserve', 'circulation:cancel_reservation',
    'reports:view', 'reports:export',
  ]),
  staff: new Set([
    'books:read',
    'members:read',
    'circulation:issue',
    'circulation:return',
  ]),
  member: new Set(),
};

function hasPermission(role, permission) {
  const granted = PERMISSIONS[role] || new Set();
  return granted.has('*') || granted.has(permission);
}

const ROLE_LABELS = {
  super_admin: 'Super Admin',
  librarian: 'Librarian',
  staff: 'Library Staff',
  member: 'Member',
};

function roleLabel(role) {
  return ROLE_LABELS[role] || role;
}

File: backend/app/permissions.py
"""
Central roles/permissions table.

Routes should never check `role == "librarian"` directly -- they ask for a
permission ("books:delete"), and this file is the only place that decides
which roles grant it. Add a role or change what Staff can touch here, and
every route that uses @permission_required(...) picks it up automatically.
"""

SUPER_ADMIN = "super_admin"
LIBRARIAN = "librarian"
STAFF = "staff"

ALL_ROLES = (SUPER_ADMIN, LIBRARIAN, STAFF)

PERMISSIONS = {
    # "*" = every permission, including settings/user management that no
    # other role gets.
    SUPER_ADMIN: {"*"},

    LIBRARIAN: {
        "books:read", "books:create", "books:update", "books:delete",
        "members:read", "members:create", "members:update", "members:delete",
        "circulation:issue", "circulation:return", "circulation:renew",
        "circulation:reserve", "circulation:cancel_reservation",
        "reports:view", "reports:export",
    },

    STAFF: {
        "books:read",
        "members:read",
        "circulation:issue",
        "circulation:return",
    },

    # Legacy self-registered patron accounts (see auth register route).
    # Deliberately granted nothing here -- self-service patrons are a
    # separate concern from staff RBAC and shouldn't silently inherit
    # access just by existing.
    "member": set(),
}


def role_has_permission(role, permission):
    granted = PERMISSIONS.get(role, set())
    return "*" in granted or permission in granted
